Mathlib/Tactic/Linarith/Preprocessing.lean

```diff
@@ -234,43 +232,42 @@ and turns it into a proof of a comparison `_ R 0`, where `R ∈ {=, ≤, <}`.
-    | _, _                                          => mkAppM ``sub_neg_of_lt #[proof]
-  | (``GE.ge, #[_, _, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
-    | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => mkAppM ``neg_nonpos_of_nonneg #[proof]
-    | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``le_zero_of_zero_ge #[proof]
-    | _, _                                          => mkAppM ``sub_nonpos_of_le #[proof]
-  | (``Not, #[a]) => do
-    let nproof ← flipNegatedComparison proof a
-    aux nproof (← inferType nproof)
-  | a => throwError m!"couldn't rearrange comparison {a}"
+    match e.getAppFnArgs with
+    | (``LE.le, #[_, _, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
+      | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => return proof
+      | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``neg_nonpos_of_nonneg #[proof]
+      | _, _                                          => mkAppM ``sub_nonpos_of_le #[proof]
+    | (``LT.lt, #[_, _, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
+      | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => return proof
+      | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``neg_neg_of_pos #[proof]
+      | _, _                                          => mkAppM ``sub_neg_of_lt #[proof]
+    | (``Eq, #[_, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
+      | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => return proof
+      | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``Eq.symm #[proof]
+      | _, _                                          => mkAppM ``sub_eq_zero_of_eq #[proof]
+    | (``GT.gt, #[_, _, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
+      | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => mkAppM ``neg_neg_of_pos #[proof]
+      | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``lt_zero_of_zero_gt #[proof]
+      | _, _                                          => mkAppM ``sub_neg_of_lt #[proof]
+    | (``GE.ge, #[_, _, a, b]) => match a.getAppFnArgs, b.getAppFnArgs with
+      | _, (``OfNat.ofNat, #[_, .lit (.natVal 0), _]) => mkAppM ``neg_nonpos_of_nonneg #[proof]
+      | (``OfNat.ofNat, #[_, .lit (.natVal 0), _]), _ => mkAppM ``le_zero_of_zero_ge #[proof]
+      | _, _                                          => mkAppM ``sub_nonpos_of_le #[proof]
+    | (``Not, #[a]) => do
+      let nproof ← flipNegatedComparison proof a
+      aux nproof (← inferType nproof)
+    | a => throwError "couldn't rearrange comparison {a}"
 
 /--
 `compWithZero h` takes a proof `h` of an equality, inequality, or negation thereof,
 and turns it into a proof of a comparison `_ R 0`, where `R ∈ {=, ≤, <}`.
  -/
-def compWithZero : Preprocessor :=
-{ name := "make comparisons with zero",
-  transform := fun e =>
-  return [← rearrangeComparison e] }
+def compWithZero : Preprocessor where
+  name := "make comparisons with zero"
+  transform e := try pure [← rearrangeComparison e] catch _ => pure []
```

Comment:
this is the actual fix for #2610

Reply:
Isn't catching everything a bit risky? Shouldn't we only catch typeclass failures?

Reply:
Okay, I've changed this to only catch synthesis exceptions, but it is probably clunky.